### ISSUE TITLE
Roslyn loses information on nullable unlifting conversions in tuple components (but tracks them if you cast it component by component)

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#51321]
**Version Used**: 
```
Branch master (9 Feb 2021)
Latest commit 2d3ffe0 by msftbot[bot]:
Merge pull request #49990 from ryzngard/feature/namespace_analyzer

Add sync namespace analyzer
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
class Program<T> where T: struct
{
  static void F1(T x)
  {
    (object, object) t = (x, x);
    ((T)t.Item1).ToString();
  }
    
  static void F2(T x)
  {
    (object, T?) t = (x, x);
    ((T?)t.Item1).Value.ToString();
  }

  static void F3(T x)
  {
    (object, object) t = (x, x);
    (((T, object))t).Item1.ToString();
  }

  static void F4(T x)
  {
    (object, object) t = (x, x);
    (((T?, object))t).Item1.Value.ToString(); // CS8629
  }
}
```

**Expected Behavior**:
No warnings in all methods

**Actual Behavior**:
`CS8629: Nullable value type may be null.` is reported for `F4` only

**Notes**
It's a bit hard to determine what behavior is expected there so I want to emphasize the following points:
- The compiler knows that `t.Item1` is not nullable in all methods (you can check it by dereferencing it)
- The compiler doesn't complain about casting the first component to non-nullable struct neither as a single element in `F1` nor as a part of a tuple in `F3`
- The compiler uses operands' state as a resulting nullable state after `T -> Nullable<T>` presumably because this cast might be required by the type system
- The compiler doesn't complain about using this component's value after casting it to `T?` in `F2` meaning that it knows that the component cannot be null even after the cast

All of them lead me to the conclusion that Roslyn should track this component's non-nulalble state through tuple conversion as well.

Note that a conversion to `(T?, anything)` tuple might be required by the type system e.g. to later pass the resulting tuple to some method which will require this exact type (e.g. by reference) or for type inference. Due to this reason it seems rather unfortunate that the compiler currently forces me to deal with these false nullability warnings if I want to do something with the first component of the resulting tuple but cannot cast it to the underlying struct for any reason.

This one might be related to other tuple issues but doesn't seem exactly the same as any of them: https://github.com/dotnet/roslyn/issues/50670, https://github.com/dotnet/roslyn/issues/47892

**Comment by TessenR** [comment]
Not sure whether this one is related but it looks very similar.
Nullable state of tuple components' members is lost on _some_ `S` <-> `Nullable<S>` conversions but not for all of them

```cs
#nullable enable
struct S
{
  internal object? F;
}
class Program
{
  static void F1()
  {
    var source = (new S() { F = 1 }, new object());
    ((S?)source.Item1).Value.F.ToString(); // all is fine
      
    var t = ((S?, object)) source;
    t.Item1.Value.F.ToString(); // all is fine
      
    (((S?, object))source).Item1.Value.F.ToString(); // all is fine
  }
    
  static void F2()
  {
    var t = ((S?, object)) (new S() { F = 1 }, new object());
    t.Item1.Value.F.ToString(); // CS8629, CS8602
      
    t = ((S?, object)) (t.Item1, new object());
    t.Item1.Value.F.ToString(); // no warnings
      
    var t2 = ((S, object)) t; // CS8619: Nullability of reference types in value of type '(S?, object)' doesn't match target type '(S, object)'.
    t2.Item1.F.ToString(); // CS8602
      
    var t3 = ((S, object)) (t.Item1, t.Item2); // no warnings
    t3.Item1.F.ToString(); // CS8602
      
    var t4 = ((S?, object)) (t.Item1, t.Item2); // no warnings
    t4.Item1.Value.F.ToString(); // no warnings
  }
}
```

**Expected behavior:**
No warnings in both methods

**Actual behavior:**
A bunch of false positive warnings reported in `F2` that is caused by lost nullability state on nullable conversions.

**Notes**
It looks like Roslyn fails to track the first component's and its members' nullability state in some cases that are rather hard to formulate. Please advise on whether I need to create a separate issue for this example

**Comment by RikkiGibson** [comment]
Does the behavior differ based on whether a `T : struct` type parameter is used versus a concrete struct type is used?

**Comment by TessenR** [comment]
@RikkiGibson It doesn't. Here's the same example where I replaced the type parameter with a concrete struct type and it has the same behavior
```cs
#nullable enable
struct T { public int X; }
class Program
{
  static void F1(T x)
  {
    (object, object) t = (x, x);
    ((T)t.Item1).ToString();
  }
    
  static void F2(T x)
  {
    (object, T?) t = (x, x);
    ((T?)t.Item1).Value.ToString();
  }

  static void F3(T x)
  {
    (object, object) t = (x, x);
    (((T, object))t).Item1.ToString();
  }

  static void F4(T x)
  {
    (object, object) t = (x, x);
    (((T?, object))t).Item1.Value.ToString(); // CS8629
  }
}
```
